pennylane/fourier/spectrum.py

```diff
@@ -0,0 +1,261 @@
+            input-encoding gates are marked by their ``id`` attribute
+        encoding_gates (list[str]): list of input-encoding gate ``id`` strings
+            for which to compute the frequency spectra
+
+    Returns:
+        (dict[str, list[float]]): Dictionary with the input-encoding gate ``id`` as keys and
+        their frequency spectra as values.
+
+
+    **Details**
+
+    A circuit that returns an expectation value which depends on
+    :math:`N` scalar inputs :math:`x_j` can be interpreted as a function
+    :math:`f: \mathbb{R}^N \rightarrow \mathbb{R}`. This function can always be
+    expressed by a Fourier-type sum
+
+    .. math::
+
+        \sum \limits_{\omega_1\in \Omega_1} \dots \sum \limits_{\omega_N \in \Omega_N}
+        c_{\omega_1,\dots, \omega_N} e^{-i x_1 \omega_1} \dots e^{-i x_N \omega_N}
+
+    over the *frequency spectra* :math:`\Omega_j \subseteq \mathbb{R},`
+    :math:`j=1,\dots,N`. Each spectrum has the property that
+    :math:`0 \in \Omega_j`, and the spectrum is
+    symmetric (for every :math:`\omega \in \Omega_j` we have that :math:`-\omega \in
+    \Omega_j`). If all frequencies are integer-valued, the Fourier sum becomes a
+    *Fourier series*.
+
+    As shown in `Vidal and Theis (2019)
+    <https://arxiv.org/abs/1901.11434>`_ and `Schuld, Sweke and Meyer (2020)
+    <https://arxiv.org/abs/2008.08605>`_, if an input :math:`x_j, j = 1 \dots N`
+    only enters into single-parameter gates of the form :math:`e^{-i x_j G}`, the
+    frequency spectrum :math:`\Omega_j` is fully determined by the eigenvalues
+    of the generators :math:`G`. In many situations, the spectra are limited
+    to a few frequencies only, which in turn limits the function class that the circuit
+    can express.
+
+    The ``spectrum`` function computes all frequencies that will potentially appear in the
+    sets :math:`\Omega_1` to :math:`\Omega_N` (which correspond to the :math:`N` different strings
+    used as an ``id`` to mark the input-encoding gates).
+
+    **Example**
+
+    Consider the following example, which uses non-trainable inputs ``x`` and
+    trainable parameters ``w`` as arguments to the qnode.
+
+    .. code-block:: python
+
+        import pennylane as qml
+        import numpy as np
+        from pennylane.fourier import spectrum
+
+        n_layers = 2
+        n_qubits = 3
+        dev = qml.device("default.qubit", wires=n_qubits)
+
+        @qml.qnode(dev)
+        def circuit(x, w):
+            for l in range(n_layers):
+                for i in range(n_qubits):
+                    qml.RX(x[i], wires=0, id="x"+str(i))
+                    qml.Rot(w[l,i,0], w[l,i,1], w[l,i,2], wires=0)
+            qml.RZ(x[0], wires=0, id="x0")
+            return qml.expval(qml.PauliZ(wires=0))
+
+        x = np.array([1, 2, 3])
+        w = np.random.random((n_layers, n_qubits, 3))
+        res = spectrum(circuit)(x, w)
+
+        for inp, freqs in res.items():
```

Comment:
Before showing the spectrum, it might be helpful to print the circuit here so that the user can see more clearly how, e.g.,  `x0` appears in the circuit 3 times, which is why the spectrum is from -3 to 3.

Reply:
Great idea!

Reply:
Only that the circuit diagram is awfully long, and a line break looks confusing.